xtask/src/ui.rs

```diff
@@ -0,0 +1,480 @@
+    //  Test help
+    //  Test version
+    //  Test images, volumes, containers, clean
+
+    Ok(())
+}
+
+// Adapted from
+// https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/tests/testsuite/support/mod.rs#L507
+fn target_dir() -> PathBuf {
+    env::current_exe()
+        .ok()
+        .map(|mut path| {
+            path.pop();
+            if path.ends_with("deps") {
+                path.pop();
+            }
+            path
+        })
+        .unwrap()
+}
+
+// Adapted from
+// https://github.com/assert-rs/assert_cmd/blob/363daac637b5f20ba3524ccf648de5f96f0dc45f/src/cargo.rs#L198-L208
+fn cargo_bin(bin: impl AsRef<str>, cwd: impl AsRef<Path>) -> cross::Result<Command> {
+    let bin = bin.as_ref();
+    let env_var = format!("CARGO_BIN_EXE_{}", bin);
+    let path = env::var_os(&env_var)
+        .map(|p| p.into())
+        .unwrap_or_else(|| target_dir().join(format!("{}{}", bin, env::consts::EXE_SUFFIX)));
+
+    match path.is_file() {
+        true => {
+            let mut cmd = Command::new(path);
+            cmd.current_dir(cwd.as_ref());
+            Ok(cmd)
+        }
+        false => eyre::bail!("unable to find cargo command {bin}"),
+    }
+}
+
+fn cross_command(
+    subcommand: &str,
+    target: &str,
+    image: &str,
+    path: impl AsRef<Path>,
+    toolchain: Option<&str>,
+) -> cross::Result<Command> {
+    let envvar = format!(
+        "CROSS_TARGET_{}_IMAGE",
+        target.replace('-', "_").to_ascii_uppercase()
+    );
+    let mut cmd = cargo_bin("cross", path)?;
+    if let Some(toolchain) = toolchain {
+        cmd.arg(&format!("+{toolchain}"));
+    }
+    cmd.args([subcommand, "--target", target]);
+    cmd.env(envvar, image);
+
+    Ok(cmd)
+}
+
+fn check_file_exists(base: impl AsRef<Path>, relpath: &str) -> cross::Result<()> {
+    let path = base.as_ref().join(relpath);
+    match path.exists() {
+        true => Ok(()),
+        false => eyre::bail!("path \"{relpath}\" unexpectedly does not exist"),
+    }
+}
+
+fn clone(url: &str, path: impl AsRef<Path>, msg_info: &mut MessageInfo) -> cross::Result<()> {
+    let mut cmd = Command::new("git");
+    let path = path.as_ref().to_utf8()?;
+    cmd.args(["clone", "--depth", "1", "--recursive", url, path]);
+    if !msg_info.is_verbose() {
+        cmd.stderr(Stdio::null());
+    }
+    cmd.run(msg_info, !msg_info.is_verbose())
+        .map_err(Into::into)
+}
+
+fn pull(engine: &docker::Engine, image: &str, msg_info: &mut MessageInfo) -> cross::Result<()> {
+    msg_info.note(format_args!(
+        "pulling container image \"{image}\": this may take a while"
+    ))?;
+
+    let mut cmd = engine.subcommand("pull");
+    cmd.arg(image);
+    if !msg_info.is_verbose() {
+        cmd.stderr(Stdio::null());
+    }
+    cmd.run(msg_info, !msg_info.is_verbose())
+        .map_err(Into::into)
+}
+
+fn build_cross(msg_info: &mut MessageInfo) -> cross::Result<()> {
+    msg_info.note("building cross: this might take a while")?;
```

Comment:
Running `xtask` doesn't necessarily build the cross binaries, so we need to run this prior. Build the full workspace, since `xtask` will benefit from it later.